File: ios/CashuWallet/Models/Mints/MintInfo.swift
import Foundation

struct MintInfo: Identifiable, Equatable, Codable {
    var id: String { url }
    let url: String
    var name: String
    var description: String?
    var isActive: Bool
    var balance: UInt64
    
    /// Icon URL (if available from mint info)
    var iconUrl: String?
    
    /// Supported units — the union of the mint's mintable and meltable units.
    var units: [String] = ["sat"]

    /// Units the mint can MINT (NUT-04), a subset of `units`. Drives the Receive
    /// unit selector so we never offer a melt-only unit for minting.
    var mintUnits: [String] = ["sat"]

    /// Supported NUT-04 payment methods for receiving
    var supportedMintMethods: [PaymentMethodKind] = [.bolt11]

    /// Supported NUT-05 payment methods for sending
    var supportedMeltMethods: [PaymentMethodKind] = [.bolt11]

    /// Nil is a legacy cache; an empty live advertisement disables that direction.
    var mintMethodSettings: [AdvertisedPaymentMethod]? = nil
    var meltMethodSettings: [AdvertisedPaymentMethod]? = nil

    func mintMethods(for unit: String) -> [PaymentMethodKind] {
        guard let settings = mintMethodSettings else { return supportedMintMethods }
        return PaymentMethodKind.ordered(settings.filter { $0.unit == unit }.map(\.method))
    }

    func methodName(_ method: PaymentMethodKind, unit: String, melt: Bool = false) -> String {
        let settings = melt ? meltMethodSettings : mintMethodSettings
        return settings?.first { $0.method == method && $0.unit == unit }?.displayName ?? method.friendlyTitle
    }

    /// NUT-04 bolt12 `MintMethodSettings.description`. Default false so records
    /// persisted before this landed, and mints that omit the field, fail closed
    /// (the Description row stays hidden until a live fetch advertises true).
    var supportsBolt12MintDescription: Bool = false

    /// Required on-chain confirmations for minting, if advertised by the mint
    var onchainMintConfirmations: Int? = nil
    
    /// Last updated timestamp
    var lastUpdated: Date = Date()
}

/// Lightweight identity + capability preview fetched for discovery / staging
/// without adding the mint to the saved wallet list.
struct MintPreviewInfo {
    let name: String?
    let iconUrl: String?
    let methods: [PaymentMethodKind]
}

extension MintInfo {
    /// True when the mint advertises more than one unit, so a unit chooser is
    /// worth surfacing. Single-unit mints hide the selector entirely.
    var supportsMultipleUnits: Bool { units.count > 1 }

    /// Preferred unit for this mint: "sat" when supported, otherwise the first
    /// unit alphabetically, falling back to "sat" for a malformed empty list.
    var defaultUnit: String {
        units.contains("sat") ? "sat" : (units.sorted().first ?? "sat")
    }

    /// Returns `unit` when this mint supports it, otherwise `defaultUnit`. Used
    /// to reset a stale selection when the active mint changes.
    func resolvedUnit(_ unit: String?) -> String {
        guard let unit, units.contains(unit) else { return defaultUnit }
        return unit
    }

    // MARK: - Mintable units (NUT-04)

    /// True when the mint can mint more than one unit — gates the Receive selector.
    var supportsMultipleMintUnits: Bool { mintUnits.count > 1 }

    /// Preferred mintable unit: "sat" when mintable, else the first sorted unit.
    var defaultMintUnit: String {
        mintUnits.contains("sat") ? "sat" : (mintUnits.sorted().first ?? "sat")
    }

    /// Returns `unit` when the mint can mint it, otherwise `defaultMintUnit`.
    func resolvedMintUnit(_ unit: String?) -> String {
        guard let unit, mintUnits.contains(unit) else { return defaultMintUnit }
        return unit
    }
}

extension MintInfo {
    private enum CodingKeys: String, CodingKey {
        case url
        case name
        case description
        case isActive
        case balance
        case iconUrl
        case units
        case mintUnits
        case supportedMintMethods
        case supportedMeltMethods
        case mintMethodSettings, meltMethodSettings
        case supportsBolt12MintDescription
        case onchainMintConfirmations
        case lastUpdated
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        url = try container.decode(String.self, forKey: .url)
        name = try container.decodeIfPresent(String.self, forKey: .name) ?? "Unknown Mint"
        description = try container.decodeIfPresent(String.self, forKey: .description)
        isActive = try container.decodeIfPresent(Bool.self, forKey: .isActive) ?? true
        balance = try container.decodeIfPresent(UInt64.self, forKey: .balance) ?? 0
        iconUrl = try container.decodeIfPresent(String.self, forKey: .iconUrl)
        units = try container.decodeIfPresent([String].self, forKey: .units) ?? ["sat"]
        // Older records predate mintUnits — fall back to the full unit set so a
        // multi-unit mint keeps offering units until its next refresh repopulates.
        mintUnits = try container.decodeIfPresent([String].self, forKey: .mintUnits) ?? units
        supportedMintMethods = try container.decodeIfPresent([PaymentMethodKind].self, forKey: .supportedMintMethods) ?? [.bolt11]
        supportedMeltMethods = try container.decodeIfPresent([PaymentMethodKind].self, forKey: .supportedMeltMethods) ?? [.bolt11]
        mintMethodSettings = try container.decodeIfPresent([AdvertisedPaymentMethod].self, forKey: .mintMethodSettings)
        meltMethodSettings = try container.decodeIfPresent([AdvertisedPaymentMethod].self, forKey: .meltMethodSettings)
        supportsBolt12MintDescription = try container.decodeIfPresent(Bool.self, forKey: .supportsBolt12MintDescription) ?? false
        onchainMintConfirmations = try container.decodeIfPresent(Int.self, forKey: .onchainMintConfirmations)
        lastUpdated = try container.decodeIfPresent(Date.self, forKey: .lastUpdated) ?? Date()
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(url, forKey: .url)
        try container.encode(name, forKey: .name)
        try container.encodeIfPresent(description, forKey: .description)
        try container.encode(isActive, forKey: .isActive)
        try container.encode(balance, forKey: .balance)
        try container.encodeIfPresent(iconUrl, forKey: .iconUrl)
        try container.encode(units, forKey: .units)
        try container.encode(mintUnits, forKey: .mintUnits)
        try container.encode(supportedMintMethods, forKey: .supportedMintMethods)
        try container.encode(supportedMeltMethods, forKey: .supportedMeltMethods)
        try container.encodeIfPresent(mintMethodSettings, forKey: .mintMethodSettings)
        try container.encodeIfPresent(meltMethodSettings, forKey: .meltMethodSettings)
        try container.encode(supportsBolt12MintDescription, forKey: .supportsBolt12MintDescription)
        try container.encodeIfPresent(onchainMintConfirmations, forKey: .onchainMintConfirmations)
        try container.encode(lastUpdated, forKey: .lastUpdated)
    }
}

// Extension for notifications

/// A method belongs to one mint, direction, and unit. Names never determine identity.
struct AdvertisedPaymentMethod: Codable, Equatable, Hashable, Identifiable {
    let method: PaymentMethodKind
    let unit: String
    var name: String? = nil
    var minAmount: UInt64? = nil
    var maxAmount: UInt64? = nil

    var id: String { "\(method.rawValue):\(unit)" }
    var displayName: String {
        guard method.isCustom else { return method.friendlyTitle }
        let label = name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return !label.isEmpty && label.count <= 30 && label.rangeOfCharacter(from: .controlCharacters) == nil
            ? label : method.displayName
    }

    func accepts(_ amount: UInt64) -> Bool {
        amount > 0 && amount >= (minAmount ?? 0) && amount <= (maxAmount ?? .max)
    }
}

File: android/app/src/main/java/com/cashu/me/Models/Mints/MintInfo.kt
package com.cashu.me.Models

import kotlinx.serialization.Serializable

@Serializable
data class MintInfo(
    val url: String,
    val name: String = "Unknown Mint",
    val description: String? = null,
    val isActive: Boolean = true,
    val balance: Long = 0,
    val iconUrl: String? = null,
    val units: List<String> = listOf("sat"),
    // NUT-04 mintable units. Empty for records stored before multi-unit landed;
    // effectiveMintUnits falls back to the full unit set until the next refresh.
    val mintUnits: List<String> = emptyList(),
    // NUT-04/NUT-05 rails, tri-state: null = never fetched (unknown — the
    // compatibility default applies), empty = the live mint reported none
    // (absent — hide the direction), non-empty = the reported methods.
    val supportedMintMethods: List<PaymentMethodKind>? = null,
    val supportedMeltMethods: List<PaymentMethodKind>? = null,
    val mintMethodSettings: List<AdvertisedPaymentMethod>? = null,
    val meltMethodSettings: List<AdvertisedPaymentMethod>? = null,
    // NUT-04 bolt12 MintMethodSettings.description. Default false so records
    // persisted before this landed, and mints that omit the field, fail closed
    // (the Description row stays hidden until a live fetch advertises true).
    val supportsBolt12MintDescription: Boolean = false,
    val onchainMintConfirmations: Int? = null,
    // NUT-06 self-reported metadata (contact / terms / software). Empty or null
    // when the mint did not report it — the UI never invents placeholders.
    val contacts: List<MintContact> = emptyList(),
    val tosUrl: String? = null,
    val software: MintSoftware? = null,
    val descriptionLong: String? = null,
    val motd: String? = null,
    val nutSupport: NutSupport = NutSupport(),
    val lastUpdatedEpochMillis: Long = System.currentTimeMillis(),
) {
    val id: String get() = url

    fun mintMethods(unit: String): List<PaymentMethodKind> = mintMethodSettings?.let { settings ->
        PaymentMethodKind.ordered(settings.filter { it.unit == unit }.map { it.method })
    } ?: effectiveMintMethods

    fun methodName(method: PaymentMethodKind, unit: String, melt: Boolean = false): String =
        (if (melt) meltMethodSettings else mintMethodSettings)
            ?.firstOrNull { it.method == method && it.unit == unit }?.displayName ?: method.friendlyTitle


    /**
     * Mint (receive) rails with the pre-fetch compatibility default: a mint whose
     * NUT-06 metadata was never fetched is assumed BOLT11; a mint that reported
     * an empty list stays empty.
     */
    val effectiveMintMethods: List<PaymentMethodKind>
        get() = supportedMintMethods ?: listOf(PaymentMethodKind.Bolt11)

    /** Melt (send) rails with the pre-fetch compatibility default (see [effectiveMintMethods]). */
    val effectiveMeltMethods: List<PaymentMethodKind>
        get() = supportedMeltMethods ?: listOf(PaymentMethodKind.Bolt11)

    val effectiveMintUnits: List<String> get() = mintUnits.ifEmpty { units }

    /** Ecash units this mint holds/advertises (send-side gating). */
    val supportsMultipleUnits: Boolean get() = units.size > 1
    val defaultUnit: String get() = defaultOf(units)
    fun resolvedUnit(unit: String?): String =
        if (unit != null && units.contains(unit)) unit else defaultUnit

    /** Units mintable over Lightning per NUT-04 (receive-side gating). */
    val supportsMultipleMintUnits: Boolean get() = effectiveMintUnits.size > 1
    val defaultMintUnit: String get() = defaultOf(effectiveMintUnits)
    fun resolvedMintUnit(unit: String?): String =
        if (unit != null && effectiveMintUnits.contains(unit)) unit else defaultMintUnit

    private fun defaultOf(candidates: List<String>): String = when {
        candidates.contains("sat") -> "sat"
        else -> candidates.sorted().firstOrNull() ?: "sat"
    }
}

/**
 * One contact channel the mint self-reports via NUT-06 (`method` is e.g.
 * "email", "nostr", "twitter"; `info` the address/handle/URL).
 */
@Serializable
data class MintContact(
    val method: String,
    val info: String,
)

/** Mint implementation name + version self-reported via NUT-06. */
@Serializable
data class MintSoftware(
    val name: String,
    val version: String,
)

/**
 * Per-NUT capability flags reported by the mint (NUT-06 info). All default false so
 * records persisted before this landed deserialize cleanly; only the live fetch
 * populates them.
 */
@Serializable
data class NutSupport(
    val tokenStateCheck: Boolean = false,    // NUT-07
    val lightningFeeReturn: Boolean = false, // NUT-08
    val restoreFromSeed: Boolean = false,    // NUT-09
    val spendingConditions: Boolean = false, // NUT-10
    val p2pk: Boolean = false,               // NUT-11
    val dleq: Boolean = false,               // NUT-12
    val htlc: Boolean = false,               // NUT-14
    val webSocket: Boolean = false,          // NUT-20
)

/** A method belongs to one mint, direction, and unit; its label is never its identity. */
@Serializable
data class AdvertisedPaymentMethod(
    val method: PaymentMethodKind,
    val unit: String,
    val name: String? = null,
    val minAmount: Long? = null,
    val maxAmount: Long? = null,
) {
    val displayName: String get() = if (!method.isCustom) method.friendlyTitle else
        name?.trim()?.takeIf { it.isNotEmpty() && it.length <= 30 && it.none(Char::isISOControl) }
            ?: method.displayName

    fun accepts(amount: Long): Boolean =
        amount > 0 && amount >= (minAmount ?: 0) && amount <= (maxAmount ?: Long.MAX_VALUE)
}
